### DISCUSSION TITLE
Some thoughts on covidHubUtils development

### DISCUSSION THREAD

**Discussion by nikosbosse** [reichlab/covidHubUtils#186] [General]
### file paths
- [x] file paths should maybe use the here package. Sam has strong opinions on that - apparently it makes things much easier when running something on a server. e.g. in the get truth functions

### documentation
- [x] documentation could maybe be more verbose to help outsiders understand it
- [x] examples in the documenation would help a lot
- [x] some of the documentation is US specific

### Function arguments
- [x] Maybe it would be clearer to set arguments that can be left blank to NULL everywhere, so that users know they don't need to provide it? This is the case in some functions, but not in all. Ideally this should be handled consistently across functions

### Zoltar connections
- [x] I think it would be good to avoid forced zoltar connections where possible. Options: 
	- [x] We could e.g. save the valid targets etc. locally? (except for the timezeros). 
	- [x] Or alternatively, make the zoltar checks optional? 

### load truth
- [x] add functionality to get different target\_variables at once
- [x] maybe add functionality for adjustments, i.e. don't give negative cases? 

### load forecasts
- [x] The load\_forecasts functions are a bit confusing in their structure and it isn't immediately clear which calls which. 
	- [x] Maybe the naming could be adapted to make that a bit clearer? 
	- [x] Maybe we could just have one load_forecasts function instead of another latest function (given that the latest function also needs an explicit forecast date)? Searching within a window also makes sense for the local hub repo in general

### additional features that would be nice
- [x] code for creating model ensembles

### Thoughts I had earlier that may be outdated and other functions I haven't looked at in detail yet
- plot\_forecast.R
	- [ ] locations are currently hard-coded

- get\_plot\_forecast\_data.R
	- [ ] fips codes hard coded

- get\_all\_models.R
	- [ ] remote\_repo option has US hard-coded
	- [ ] zoltar option has US hard-coded
	- [ ] some more documentation for what it actually does might be helpful. E.g.: where does the path to the local repo have to point to? 
	
- get\_hub\_locations.R
	- [ ] has US locations hard-coded
	- [ ] current European file is here: https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/main/template/locations\_eu.csv and doesn't have abbreviations - so maybe the formatting would also have to be changed
	
- [ ] get\_model\_designations.R
	- [ ] documentations is specific to US hub
	- [ ] not sure, but I think it can't handle spaces in your file names? 

- get\_truth.R
	- [ ] unsure whether this is only used by maintainers? 
	- [ ] preprocess\_truth\_for\_zoltar has only deaths and that is hard-coded. If we want to use it it probably needs some adaptions. 

- date\_management.R
	- [ ] maybe some documentation

**Comment by elray1** [comment]
r.e. file paths in the functions get_truth.R -- I agree that using the here package is useful, but I think that use of here belongs in the script calling those functions rather than in the functions themselves

**Comment by nickreich** [comment]
Re: documentation- we added issues #192 #193 #194 #195 to improve documentation.

**Reply by elray1** [reply]
Nikos, do you have additional specific suggestions for places where documentation could be more verbose?

**Comment by elray1** [comment]
r.e.: "maybe add functionality for adjustments, i.e. don't give negative cases?"

For now, we've decided adjusting truth data is out of scope.  There is some functionality related to this in the covidData package, and we're working towards adding support for European hub truth data there.

**Comment by nickreich** [comment]
Issue #198 created to have a designated place to iterate on re-design of load_forecasts.

**Comment by nickreich** [comment]
We would happily accept a PR to create simple ensembles. given our existing functionality in [covidEnsembles](https://github.com/reichlab/covidEnsembles), we are not going to prioritize this development.
